File: scala/el-entity-mapper/src/main/scala/net/freedom/gj/mapper/config/SimpleXmlMapperConfiguration.scala
package net.freedom.gj.mapper.config

import scala.collection.mutable._
import scala.io.Source
import scala.xml.NodeSeq
import scala.xml.parsing.ConstructingParser

class SimpleXmlMapperConfig
(val file: String, val sourceType: String, val targetType: String) extends MapperConfigTrait {

  lazy val mi: MappingInformation = parse()

  override def getMappingInformation: MappingInformation = mi

  private def parse(): MappingInformation = {
    val resource = classOf[SimpleXmlMapperConfig].getResourceAsStream(file);
    val parser = ConstructingParser.fromSource(Source.fromInputStream(resource), false);
    val doc = parser.document()

    var list = new ListBuffer[MappingData]();
    for (element <- doc \ "bind") {
      list += getBindData(element)
    }
    for (element <- doc \ "collection") {
      var b = new ListBuffer[MappingData]()
      element.child.foreach(bind => b += getBindData(bind))
      list += new MappingData((element \ "@source").text,
        (element \ "@target").text, null, b, (element \ "@type").text)
    }
    val processorList= new ListBuffer[PostProcessor]()
    for (processors <- doc \ "post-processors") {
      processors.child.foreach(p => processorList += newInstance((p \ "@value").text, classOf[PostProcessor]))
    }
    new MappingInformation(list, processorList)
  }

  private def getBindData(el: NodeSeq): MappingData = {
    new MappingData((el \ "@source").text,
      (el \ "@target").text,
      getConverter((el \ "@converter").text))
  }
}

File: root/entity-mapper/src/main/java/net/freedom/gj/beans/config/SimpleXmlMapperConfiguration.java
/*
 * Copyright 2009-2010 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.freedom.gj.beans.config;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;
import net.freedom.gj.beans.mapper.MappingData;
import net.freedom.gj.beans.mapper.MappingInformation;
import net.freedom.gj.beans.mapper.PostProcessor;
import net.freedom.gj.beans.util.Lg;
import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

/**
 *
 * @author Martin Jamszolik
 */
public class SimpleXmlMapperConfiguration extends AbstractMapperConfiguration {

   
    public MappingInformation getMappingInformation() {
        try {
            if (mappingInformation != null) {
                return mappingInformation;
            }
            mappingInformation = new MappingInformation();
            InputStream stream = this.getClass().getResourceAsStream(configurationFile);            
            
            SAXParserFactory spf = SAXParserFactory.newInstance();
            SAXParser sp = spf.newSAXParser();
            sp.parse(stream, new XmlDefaultHandler(mappingInformation));
            stream.close();
        } catch (Exception ex) {
            Lg.log(this, Lg.ERROR, "Error", ex);
        }
        return mappingInformation;
    }

    

    class XmlDefaultHandler extends DefaultHandler {

        private MappingInformation mappingInformation;

        private MappingData collectionData;
        private List<PostProcessor> procList;

        public XmlDefaultHandler(MappingInformation info) {
            mappingInformation = info;
        }

        @Override
        public void endElement(String uri, String localName, String qName) throws SAXException {
           if( qName.equals("collection") ){
               mappingInformation.getMappingData().add(collectionData);
               collectionData = null;
           }else if( qName.equals("post-processors") ){
               mappingInformation.setPostProcessors(procList);
               procList = null;
           }
        }
        @Override
        public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
            if (qName.equals("bind")) {
                MappingData d = new MappingData();
                d.setSourceExpression(attributes.getValue("source"));
                d.setTargetExpression(attributes.getValue("target"));
                d.setConverter(getConverter( attributes.getValue("converter")  ));
                
                if( collectionData != null )
                    collectionData.getCollectionMappingData().add(d);
                else
                    mappingInformation.getMappingData().add(d);

            }else if( qName.equals("collection") ){
                collectionData = new MappingData();
                collectionData.setCollection(true);
                collectionData.setSourceExpression(attributes.getValue("source"));
                collectionData.setTargetExpression(attributes.getValue("target"));
                collectionData.setCollectionObjectType(attributes.getValue("type"));
                collectionData.setCollectionMappingData(new ArrayList<MappingData>());

            }else if ( qName.equals("post-processors") ){
                procList = new ArrayList<PostProcessor>();

            }else if( qName.equals("processor") ){
                PostProcessor proc = (PostProcessor)getInstance(attributes.getValue("value") );
                if( proc != null )
                procList.add( proc );
            }
        }

        private Object getInstance(String impl){
            try {
                return (Class.forName(impl)).newInstance();
            } catch (Exception ex) {
                Lg.log(Lg.ERROR,"Faild to create new Instance of {0}. Error: {1}",impl,ex.getMessage() );
                return null;
            }
        }
    }

    
}
